### ISSUE TITLE
Understanding train/test/split 

### ISSUE THREAD

**Issue by slavakx** [facebookexperimental/Robyn#735]
My question is related to #646 

I got the logic of finding the optimal training size. But I have several questions because I coudln't figure out answers in your code: 

- are you using the test set metric for optimization in any way? 
- If not, why do you split into train/test/val and not say train/val with varying train size?
- correct me if I am wrong: the final model is the model built on the train set (excluding test and validation) and not the one that is retrained on all the data after best parameters were found? If yes, what is the idea behind it

Thanks!

**Comment by gufengzhou** [comment]
- When setting ts_validation = TRUE, the nrmse_val (from validation set) is used as objective function in the optimisation. The 3 way split is very common in machine learning, esp. for hyperparameter optimisation, IMO.

**Comment by slavakx** [comment]
It should be probably moved to discussion. 
I have several points I'd like you to address for better understanding your decision making regarding cross validation

> There's no "retrain on all data with best parameters", because defining the parameters is the purpose of training itself.
Yes and no. When the data is not a time-series data, one may train the final model on a subset of the data (train set) or refit the data with the best parameters on full data. In case of time-series data, one must retrain on most recent data to perform next step prediction (next week prediction and so on). Since Robyn performs trend and seasonality decomposition prior to modeling, not retraining on full data is indeed not critical. 

> defining the parameters is the purpose of training itself
True. But what if I apply a time-series cross validation for better generalizability instead of just one train/test split?
Let's say I apply 5 week (fold) cross validation, training 5 models with different train length. 
This is very common in ML too. So in this case I would use the best parameters to train my model on most recent data.

**Comment by gufengzhou** [comment]
> Yes and no. When the data is not a time-series data, one may train the final model on a subset of the data (train set) or refit the data with the best parameters on full data. In case of time-series data, one must retrain on most recent data to perform next step prediction (next week prediction and so on). Since Robyn performs trend and seasonality decomposition prior to modeling, not retraining on full data is indeed not critical.

"Refit the data with the best parameters on full data" is not fitting a model, it's just prediction. The purpose of model fitting is to obtain estimates for those parameters. When the parameters are selected, there's no fitting anymore.

Regarding time series cross validation, we decided against it because it will heavily impact the running time. 5-fold means almost 5x iterations, a strong trade-off we don't want to take. Moreover, cross-validation doesn't solve the "training size definition" issue, it only provides test statistics. With our evolutionary approach, we also automate the training size definition and thus reduce human bias.
